### PR TITLE
[Nullability Annotations to Java Classes] Add Missing Nullability Annotations to All SDK Override `onNewIntent(...)` Methods (`risky`)

### PR DESCRIPTION
**Pull request by ParaskP7**
Parent #18909

This PR adds the missing `@NonNull` annotation to all SDK override `onNewIntent(...)` related method on `Activity` classes, including updating any Kotlin related classes that had that as `@Nullable` via the use of the `Intent?` declaration.

FYI: This change is `risky`, meaning that there are compile-time changes associated with this change and thus needs testing to verify correctness.

-----

Nullability Annotation List:

1. [Add missing nn-a to on new intent methods for activities](https://github.com/wordpress-mobile/WordPress-Android/commit/2adc6dbbfaeb4f354e46566b277545d7f28e2d3f)

Lint Warnings Suppression List:

1. [Suppress non constant res id lint warning for media browser](https://github.com/wordpress-mobile/WordPress-Android/commit/e65f780ff0f755c3bc1ad62438efb141c8824765)

Risky List:

1. [Make intent non null on the on new intent functions for kotlin](https://github.com/wordpress-mobile/WordPress-Android/commit/bd62ee79c2ca4c54bc0d0b1a8e4bfb42a1e9afa7)

-----

## To test:

- Smoke test any `onNewIntent` related functionality on both, the WordPress and Jetpack apps, and see if everything is working as expected.
- In addition to the above smoke test, you can expand the below and follow the inner and more explicitly test steps within, which is mainly related to the risky changes that are associated with this PR:

<details>
    <summary>1. Stats Screen [StatsActivity.kt]</summary>

ℹ️ This test applies to the `Jetpack` app.

- Go to `Stats` screen and verify everything works as expected.
- Then, add any `Stats` related widget, like the `All-time` widget, configure and place it.
- Finally, click on that `Stats` widget you just added and via `debug` mode, check the `onNewIntent(...)` function is called and the `Intent` is not `null`.

</details>

<details>
    <summary>2. Scan Screen [ScanActivity.kt + ScanFragment.kt]</summary>

ℹ️ This test applies to the `Jetpack` app.

- Go to `Scan` screen and verify everything works as expected.
- Then, add a new `Threat` to your site, like the `Vulnerable Plugin: elementor (version 3.7.7)` and run a new scan. This new `Threat` should be found and appear on the main screen.
- Finally, click on that `Threat` you just added then fix it. Via `debug` mode, check the `onNewIntent(...)` function is called and the `Intent` is not `null`.

</details>

<details>
    <summary>3. Pages Screen [PagesActivity.kt]</summary>

ℹ️ This test applies to both, the `Jetpack` and `WordPress` app.
⁉️ This flow might not longer be valid as I couldn't make it work.

- Go to `Pages` screen and verify everything works as expected.
- Then, open a `Page` and edit it. Quickly exist the `Pages` screen before the upload is finished.
- Finally, when the `Page` upload is done and the notification is shown, click on that `Notification` and via `debug` mode, check the `onNewIntent(...)` function is called and the `Intent` is not `null`.

FYI: I got the test instruction above from [this](https://github.com/wordpress-mobile/WordPress-Android/pull/8327) PR.
PS: No matter, this test instruction is just informational-only as the `onNewIntent(...)` signature of this `PagesActivity.kt` class was already using a non-null `intent`, with no changes happening here.

</details>

<details>
    <summary>4. Posts Screen [PostsListActivity.kt]</summary>

ℹ️ This test applies to both, the `Jetpack` and `WordPress` app.
⁉️ This flow might not longer be valid as I couldn't make it work.

- Got to `Posts` screen and verify everything works as expected.
- Then, turn off WiFi, create a new `Post` and publish it. Switch site, open the `Posts` screen on that site and turn the WiFi back on.
- Finally, when the `Post` upload is done and the notification is shown, click on that `Notification` and via `debug` mode, check the `onNewIntent(...)` function is called and the `Intent` is not `null`.

FYI: I got the test instruction above from [this](https://github.com/wordpress-mobile/WordPress-Android/pull/9290) PR.
PS: No matter, this test instruction is just informational-only as the `onNewIntent(...)` signature of this `PostsListActivity.kt` class was already using a non-null `intent`, with no changes happening here.

</details>

-----

## Regression Notes

1. Potential unintended areas of impact

    - Potential new NPE start happening on the `Stats` and/or `Scan` screens, always related to the `onNewIntent(...)` functionality, which had its 'intent' parameter updated from being a nullable one to a non-null.

3. What I did to test those areas of impact (or what existing automated tests I relied on)

    - See `To test` section above.

4. What automated tests I added (or what prevented me from doing so)

    - N/A

-----

## PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:

- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/StatsActivity.kt [hunk 1]
```diff
@@ -50,12 +50,10 @@ class StatsActivity : LocaleAwareActivity() {
         return super.onOptionsItemSelected(item)
     }
 
-    override fun onNewIntent(intent: Intent?) {
-        intent?.let {
-            val siteId = intent.getIntExtra(WordPress.LOCAL_SITE_ID, -1)
-            if (siteId > -1) {
-                viewModel.start(intent, restart = true)
-            }
+    override fun onNewIntent(intent: Intent) {
```
Comment (mkevins) [inline WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/StatsActivity.kt:53]: I'm not sure why, but I'm getting a warning in Android Studio 🤔 :

```
Overriding method should call `super.onNewIntent`
```

However, the breakpoint was reached, and the `Intent` was non-null 🤷‍♂️ , and I don't think this is an issue with this PR.

**Review by mkevins — COMMENTED** [review]
Thanks for working on this Petros!

> meaning that there are compile-time changes associated with this change 

I assume you meant "run-time" for these "risky" changes, and not "compile-time", but let me know if that assumption is invalid 😅 .

I tested the first scenario, and it is working as expected in my case 👍 . I also was unable to test scenarios 3 and 4, like you mentioned in the description 🤷‍♂️ .

For scenario 2, I'm not sure why, but I don't see the "Scan" option. Is there perhaps some pre-requisite for getting it to appear? I was using a Business site in the Jetpack app, but did not see the option.

That said, I think these changes are ready to merge, but before doing so, I think we should creating a tracking issue (and it could be added to both the nullability and crash focus project boards). This will make it easy to keep track pro-actively, and have some tasks there to make rotation release wranglers and beta-testers aware when it lands, in case we introduce an NPE. In this case, I don't _think_ we will, but this PR could serve as a lower-risk example for establishing the pattern we use for later PRs with a larger surface area of change. Wdyt?

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/StatsActivity.kt (hunk 1 above)
Comment (ParaskP7) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/StatsActivity.kt:53]: Yeaaa @mkevins , I think we already (briefly) discussed about that. From the looks of it, to my understanding, since this is also happening on `trunk`, this an IDE bug and I have already provided feedback on this warning, that this is a `False Positive`. I am not sure what else we can do here, or why this warning is shown on some of those `onNewIntent(...)` override methods and not on some others... 🤷

FYI: Personally, I chose to ignore this for now. ☺️

**Comment by ParaskP7** [comment]
Thanks for reviewing and testing this @mkevins , much appreciated! 🙇 ❤️ 🚀 

> I assume you meant "run-time" for these "risky" changes, and not "compile-time", but let me know if that assumption is invalid 😅 .

Actually, and as discussed already, I did mean `compile-time`, not `run-time`. 🤔 

I am not sure if I understood it wrong, but we decided that any nullability related PR that introduces a `compile-time` change, that is, in addition to just adding a nullability annotation, especially a branching change such as here, which removes the need for a null check/guard, in this case removing the need to guard the Kotlin code with `?.let { ... }` (see [here](https://github.com/wordpress-mobile/WordPress-Android/pull/19100/files#diff-7bbb22a894748ee07f06e27ee38f96eeb052cb5520be5d306efe157f019ab50eL36), [here](https://github.com/wordpress-mobile/WordPress-Android/pull/19100/files#diff-04009d492a16bb1990425f0c90eb87fb533ea0cde8dc12d919c6de5f5e29efbfL196) and [here](https://github.com/wordpress-mobile/WordPress-Android/pull/19100/files#diff-6133f8675b0aefbf042ecf5d6226d83094e70278bf0b2502f7134ce9e822b012L54)), should be considered `risky` as it may not produce a NPE, whereas it wouldn't previously.

I might totally got this wrong, can you help clarify that again? 🙏 😊 🙏 

> I tested the first scenario, and it is working as expected in my case 👍 . I also was unable to test scenarios 3 and 4, like you mentioned in the description 🤷‍♂️ .

🤷 🤷 🤷 

> For scenario 2, I'm not sure why, but I don't see the "Scan" option. Is there perhaps some pre-requisite for getting it to appear? I was using a Business site in the Jetpack app, but did not see the option.

Yeaaa, to be honest, to be able to (a.) have the `Scan` option appear on your site, and then (b.) to add a vulnerability to your site and then run a new scan, it would take you some time to configure. Thus, just to keep it simple, I didn't specify those instructions, leaving it for the reviewer's to decide whether they would like to test that flow as well.

To answer you question about (a.), you could take a look at this [SiteUtils.isScanEnabled(...)](https://github.com/wordpress-mobile/WordPress-Android/blob/0c832be0df4b5f88f99fb8c79aebb24d4f4f40d6/WordPress/src/main/java/org/wordpress/android/util/SiteUtils.java#L380-L382) method, where amongst other conditions, it specifies that scan needs to be purchases on your site, maybe that's what you need to do. 🤔

To answer your next question about (b.), that is, if you still want to test that, you could check the test prerequisite on PRs such as [this](https://github.com/wordpress-mobile/WordPress-Android/pull/13768), where it guides you to download, add and install a threat (bad calendar plugin). 🍀

FYI: If that helps, I personally tested this scenario already. ☺️ 

> That said, I think these changes are ready to merge, but before doing so, I think we should creating a tracking issue (and it could be added to both the nullability and crash focus project boards).

Great, I agree, could you create one for this PR, add that then to both the nullability and crash focus project boards, and then let me know about it. I am not sure how you would like a tracking issue to look like and you helping me on that would be very much appreciated. 🙏 

> This will make it easy to keep track pro-actively, and have some tasks there to make rotation release wranglers and beta-testers aware when it lands, in case we introduce an NPE.

Totally! 💯 🥇

Could you write down some steps on how this would look like, like a (1.), (2.), (3.) step process and based on the tracking issue above. 🙏

> In this case, I don't think we will, but this PR could serve as a lower-risk example for establishing the pattern we use for later PRs with a larger surface area of change. Wdyt?

Yes, agree, you never know, I too don't think we will, but better be safe(r) than sorry. 💯 🥇

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/StatsActivity.kt (hunk 1 above)
Comment (mkevins) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/StatsActivity.kt:53]: > I think we already (briefly) discussed about that.

I recall discussing a similar issue, but I don't remember encountering this specific issue before. 🤷‍♂️  In any case, if it's on `trunk`, it's certainly not related with this PR (as I already suspected).

**Comment by mkevins** [comment]
> Actually, and as discussed already, I did mean compile-time, not run-time. 🤔

Since we are "guessing" that the intent is non-null, this PR's strategy _does_ affect run-time behavior, not only compile-time. For example, in [this change](https://github.com/wordpress-mobile/WordPress-Android/pull/19100/files#diff-04009d492a16bb1990425f0c90eb87fb533ea0cde8dc12d919c6de5f5e29efbfR195-R201), we remove a `?.let{}` guard, instead failing with `NPE` if intent turns out to be `null`, whereas previously, this would not have occurred.

> as it may not produce a NPE, whereas it wouldn't previously.

I assume here you meant: "as it _may_ <s>not</s> produce a NPE, whereas it wouldn't previously.", and if so, I think we are on the same page. To clarify, I consider that introducing an additional chance of throwing an exception where we would have otherwise failed silently is a change in run-time behavior (hence why we've labeled these as "risky", in case our guess is wrong).

More generally, the impetus for caution (and cause for pro-active monitoring) is that the framework explicitly does _not_ guarantee the nullity of these values.

> FYI: If that helps, I personally tested this scenario already.

That does help, and I agree that the elaborate setup required for some of these tests is neither necessary nor sufficient, as even if everything goes smooth in our tests, it can't rule out possible NPEs in production, where various edge cases are eventually encountered. So, I agree with putting more emphasis on monitoring during beta and production release.

> Great, I agree, could you create one for this PR, add that then to both the nullability and crash focus project boards, and then let me know about it. I am not sure how you would like a tracking issue to look like and you helping me on that would be very much appreciated. 

Sure, I'll open a task for this on the board 👍 .

> Could you write down some steps on how this would look like, like a (1.), (2.), (3.) step process and based on the tracking issue above. 

Indeed, I'll add them to the task, and I'll ping you when I do.

**Comment by ParaskP7** [comment]
👋 @mkevins !

> Since we are "guessing" that the intent is non-null, this PR's strategy does affect run-time behavior, not only compile-time. For example, in [this change](https://github.com/wordpress-mobile/WordPress-Android/pull/19100/files#diff-04009d492a16bb1990425f0c90eb87fb533ea0cde8dc12d919c6de5f5e29efbfR195-R201), we remove a ?.let{} guard, instead failing with NPE if intent turns out to be null, whereas previously, this would not have occurred.

Ah, for sure, I think we are on the same page here. 💯

So, when I am talking about a PR being `risky`, it is because it has compile-time change, which of course also translate to run-time change in behavior. That's why I am not talking about the run-time change explicitly. On the other hand, when a PR has no compile-time changes (see #19096), then it for sure doesn't have run-time changes, thus it is `safe`.

Does this make sense to you? Would you still prefer if I talk about run-time changes explicitly as well? 🤔

> I assume here you meant: "as it may not produce a NPE, whereas it wouldn't previously.", and if so, I think we are on the same page.

Ah yea, I went and corrected the `not` with `now`. 😭

FYI: I keep doing these kind of mistakes from time to time, even if I re-read the text one or more time, `now` -> `not`, `sign` -> `sing`, etc. Me being dyslexic doesn't help much, apologies and thanks for getting the context right anyway... 😅 🙏

> To clarify, I consider that introducing an additional chance of throwing an exception where we would have otherwise failed silently is a change in run-time behavior (hence why we've labeled these as "risky", in case our guess is wrong).

👍 

> More generally, the impetus for caution (and cause for pro-active monitoring) is that the framework explicitly does not guarantee the nullity of these values.

👍 

> That does help, and I agree that the elaborate setup required for some of these tests is neither necessary nor sufficient, as even if everything goes smooth in our tests, it can't rule out possible NPEs in production, where various edge cases are eventually encountered. So, I agree with putting more emphasis on monitoring during beta and production release.

👍 

> Sure, I'll open a task for this on the board 👍 .

Thanks you! 🙇 

> Indeed, I'll add them to the task, and I'll ping you when I do.

Will much appreciate the ping, thanks! 🙏

**Comment by Gilbert1874** [comment]
Is there something wrong with my new phone,

**Comment by mkevins** [comment]
> Does this make sense to you? Would you still prefer if I talk about run-time changes explicitly as well?

I think I understand, and perhaps we were using these terms _slightly_ differently. From what you wrote, I think you mean by "compile-time" changes that the binary output produced by compilation is different, and thus, could possibly lead to "run-time" changes.


To elaborate on what I was meaning by this, my use of those terms was more to do with when _errors_ are surfaced (e.g. at "compile-time", or at "run-time"). And, though we'd ideally want as many errors surfaced at compile-time as possible, certain ambiguities (notably, the ambiguous nullity of framework methods) force us to handle the possibility of `null` at run-time. In some cases, we already have a guard in place, preventing NPEs, and in other cases, we expose ourselves to NPEs implicitly, by our unguarded usage.

The defensive programming strategy is perfectly valid in many circumstances (e.g. let guard), and should not be assumed to be "failing silently", as it may not even be a failure at all (e.g. there are styles of programming in which `null` is semantically useful). In such cases, removing the guard and failing fast is a regression - hence the labeling as "risky". On the other hand, if the only changes in a PR are to do with the annotations themselves, and compiler enforcement (without any logical changes), the run-time behavior would not be altered.

Technically 🤓 , afaik, the retention on the `androidx.annotation`s `@NonNull` and `@Nullable` is set as `BINARY`, so such changes (in the "safe" PRs) may indeed alter the binary output of the build, but are invisible to reflection, and thus should not, under normal circumstances, have any affect on run-time behavior.

> Me being dyslexic doesn't help much, apologies

No worries at all, and no need to apologize! I also make these kinds of typos, and even after proof-reading fail to spot them. I still got your meaning, but just wanted to confirm rather than assume. 😃

**Comment by ParaskP7** [comment]
👋 @mkevins !

> I think I understand, and perhaps we were using these terms slightly differently. From what you wrote, I think you mean by "compile-time" changes that the binary output produced by compilation is different, and thus, could possibly lead to "run-time" changes.

Exactly, and to be honest, got and liked that idea from you. 💯 ☺️ 🙇 

Thus, I am now suffixing all such PRs with `safe`, `relatively safe`, `risky` and `breaking`, just to give a quick overview to help the review understand what to expect, plus, providing a quick description on what that `characterization` mean in the context of that specific PR!

> To elaborate on what I was meaning by this, my use of those terms was more to do with when errors are surfaced (e.g. at "compile-time", or at "run-time"). And, though we'd ideally want as many errors surfaced at compile-time as possible, certain ambiguities (notably, the ambiguous nullity of framework methods) force us to handle the possibility of null at run-time. In some cases, we already have a guard in place, preventing NPEs, and in other cases, we expose ourselves to NPEs implicitly, by our unguarded usage.

👍 

> The defensive programming strategy is perfectly valid in many circumstances (e.g. let guard), and should not be assumed to be "failing silently", as it may not even be a failure at all (e.g. there are styles of programming in which null is semantically useful). In such cases, removing the guard and failing fast is a regression - hence the labeling as "risky". On the other hand, if the only changes in a PR are to do with the annotations themselves, and compiler enforcement (without any logical changes), the run-time behavior would not be altered.

👍 

> Technically 🤓 , afaik, the retention on the androidx.annotations @NonNull and @Nullable is set as BINARY, so such changes (in the "safe" PRs) may indeed alter the binary output of the build, but are invisible to reflection, and thus should not, under normal circumstances, have any affect on run-time behavior.

🤓

> No worries at all, and no need to apologize! I also make these kinds of typos, and even after proof-reading fail to spot them. I still got your meaning, but just wanted to confirm rather than assume. 😃

🙇

**Comment by ParaskP7** [comment]
👋 @mkevins !

> Sure, I'll open a task for this on the board 👍 .
> ...
> Indeed, I'll add them to the task, and I'll ping you when I do.

A friendly ping on [these two actions above](https://github.com/wordpress-mobile/WordPress-Android/pull/19100#issuecomment-1726886231) from your side so we can move forward with this PR. 🙏

**Comment by mkevins** [comment]
I've added an issue for tracking this [here](https://github.com/wordpress-mobile/WordPress-Android/issues/19305).

**Review by mkevins — APPROVED** [review]
Now that we have a tracking issue, this is ready to merge. We can make a note of the release it lands in so that we can relay that to release wranglers and beta testers, as described in the issue.

The issue is written generically, with the intention that this should gradually scale to larger change-sets if things go smoothly.

**Comment by ParaskP7** [comment]
👋 @mkevins !

> I've added an issue for tracking this https://github.com/wordpress-mobile/WordPress-Android/issues/19305.

Awesome, thanks so much, this issue looks great! 🥇 

> Now that we have a tracking issue, this is ready to merge.

🙇 ❤️ 🚀 

> We can make a note of the release it lands in so that we can relay that to release wranglers and beta testers, as described in the issue.

I added the `23.5` milestone, did the same on #19305 and updated the status of both its projects to `In Progress`, will keep an eye on this when `23.5` gets code-freezed the Monday after the next one (`16th of October`).

> The issue is written generically, with the intention that this should gradually scale to larger change-sets if things go smoothly.

Once more, thanks for reviewing this and helping making sure that we are not introducing any unnecessary regressions in production! 💯
